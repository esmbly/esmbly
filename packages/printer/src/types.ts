export type Callback = (error: Error | null | undefined) => void;

export type Encoding = string | undefined;

export type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;
