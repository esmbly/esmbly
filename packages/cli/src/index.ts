import yargs, { Arguments } from 'yargs';
import * as init from './commands/init';
import * as run from './commands/run';

export function runCLI(): Arguments {
  return yargs
    .scriptName('esmbly')
    .command(init)
    .command(run)
    .help().argv;
}
