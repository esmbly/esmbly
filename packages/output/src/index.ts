type Callback = (error: Error | null | undefined) => void;
type Encoding = string | undefined;
type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;

class Output {
  private stdout: NodeJS.WriteStream | null;

  private stderr: NodeJS.WriteStream | null;

  public constructor() {
    this.stdout = process ? process.stdout : null;
    this.stderr = process ? process.stderr : null;
  }

  public get isTTY(): boolean {
    if (this.stdout) {
      return this.stdout.isTTY === true;
    }
    return false;
  }

  public err(chunk: string, encoding?: Encoding, callback?: Callback): boolean {
    if (this.stderr) {
      return this.stderr.write(chunk, encoding, callback);
    }
    return false;
  }

  public out(chunk: string, encoding?: Encoding, callback?: Callback): boolean {
    if (this.stdout) {
      return this.stdout.write(chunk, encoding, callback);
    }
    return false;
  }

  public clearLine(...args: WriteFn[]): void {
    if (this.isTTY) {
      args.forEach(fn => fn.call(this, '\x1b[999D\x1b[K'));
    }
  }

  public setErrorStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stderr = stream as NodeJS.WriteStream;
  }

  public setOutStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stdout = stream as NodeJS.WriteStream;
  }
}

export default new Output();
