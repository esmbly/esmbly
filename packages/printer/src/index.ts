import { Warning } from '@esmbly/types';
import chalk from 'chalk';
import { Callback, Encoding, WriteFn } from './types';
import * as templates from './templates';

class Printer {
  private stdout: NodeJS.WriteStream | null;
  private stderr: NodeJS.WriteStream | null;

  public constructor() {
    this.stdout = process ? process.stdout : null;
    this.stderr = process ? process.stderr : null;
  }

  public get isTTY(): boolean {
    if (this.stdout) {
      return this.stdout.isTTY === true;
    }

    return false;
  }

  public error(
    chunk: string,
    encoding?: Encoding,
    callback?: Callback,
  ): boolean {
    if (this.stderr) {
      return this.stderr.write(chunk, encoding, callback);
    }

    return false;
  }

  public print(
    chunk: string,
    encoding?: Encoding,
    callback?: Callback,
  ): boolean {
    if (this.stdout) {
      return this.stdout.write(chunk, encoding, callback);
    }

    return false;
  }

  public printWarnings(warnings: Warning[]): void {
    if (this.stdout) {
      warnings.forEach((warning: Warning) => {
        this.print(templates.warning(warning));
      });
    }
  }

  public forceEnableColors(): void {
    chalk.enabled = true;
  }

  public forceDisableColors(): void {
    chalk.enabled = false;
  }

  public clearLine(...args: WriteFn[]): void {
    if (this.isTTY) {
      args.forEach(fn => fn.call(this, '\x1b[999D\x1b[K'));
    }
  }

  public setErrorStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stderr = stream as NodeJS.WriteStream;
  }

  public setOutStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stdout = stream as NodeJS.WriteStream;
  }
}

export const printer = new Printer();
export { templates };
