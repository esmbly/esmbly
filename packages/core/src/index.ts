import { File, RunConfig } from '@esmbly/types';
import { parse } from './parse';
import { validateConfig, validateInputFormat } from './validate';

export async function run(config: RunConfig): Promise<File[]> {
  validateConfig(config);

  let files: File[] = [];
  const trees = parse(config.input, config.transformers[0]);

  for (const transformer of config.transformers) {
    if (transformer.before) {
      await transformer.before();
    }

    validateInputFormat(trees, transformer);

    if (transformer.transform) {
      await transformer.transform(trees);
    }

    if (transformer.createFiles) {
      files = [
        ...files,
        ...(await transformer.createFiles(trees, config.output)),
      ];
    }

    if (transformer.after) {
      await transformer.after();
    }
  }

  return files;
}
