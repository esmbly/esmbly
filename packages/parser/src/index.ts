import parse from './parse';

export default { parse };
