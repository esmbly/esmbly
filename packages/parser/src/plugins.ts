import { ParserPlugin } from '@babel/parser';
import { FileType } from '@esmbly/types';

export function getPluginsForFileType(fileType: FileType): ParserPlugin[] {
  switch (fileType) {
    case FileType.JavaScript:
      return [];
    case FileType.Flow:
      return ['flow', 'flowComments'];
    case FileType.TypeScript:
      return ['typescript'];
    default:
      throw new Error(`File type ${fileType} is not supported`);
  }
}
