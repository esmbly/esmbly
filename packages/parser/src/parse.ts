import * as parser from '@babel/parser';
import recast from 'recast';
import { AST, File } from '@esmbly/types';
import { getPluginsForFileType } from './plugins';

export default function parse(files: File[]): AST[] {
  return files.map((file: File) => {
    const tree = recast.parse(file.content, {
      parser: {
        parse(source: string) {
          return parser.parse(source, {
            sourceType: 'module',
            plugins: getPluginsForFileType(file.type),
          });
        },
      },
    });
    const ast: AST = {
      name: file.name,
      path: file.path,
      type: file.type,
      tree: tree,
      toFile: () => ({
        ...file,
        content: recast.print(tree).code,
      }),
    };
    return ast;
  });
}
