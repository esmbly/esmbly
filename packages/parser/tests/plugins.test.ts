import { FileType } from '@esmbly/types';
import { getPluginsForFileType } from '../src/plugins';

describe('getPluginsForFileType', () => {
  it('returns the correct plugins for JavaScript files', () => {
    const plugins = getPluginsForFileType(FileType.JavaScript);
    expect(plugins).toEqual(['flow', 'flowComments']);
  });

  it('returns the correct plugins for TypeScript files', () => {
    const plugins = getPluginsForFileType(FileType.TypeScript);
    expect(plugins).toEqual(['typescript']);
  });

  it('throws an error for file types that are not yet supported', () => {
    expect(() => {
      getPluginsForFileType(FileType.WebAssembly);
    }).toThrowErrorMatchingSnapshot();
  });
});
