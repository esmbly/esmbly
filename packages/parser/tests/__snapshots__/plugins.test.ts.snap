// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPluginsForFileType throws an error for file types that are not yet supported 1`] = `"File type WebAssembly is not supported"`;
